packages/node_modules/@ciscospark/internal-plugin-metrics/src/metrics.js

```diff
@@ -99,6 +114,107 @@ const Metrics = SparkPlugin.extend({
       eventPayload: payload
     };
     return this.callDiagnosticEventsBatcher.request(event);
+  },
+
+  storeMetric(id, metricProps, metricType) {
+    const tuple = {metric: metricProps, type: metricType};
+    return this.metricMap.set(id, tuple);
+  },
+
+  getMetrics() {
+    return this.metricMap;
+  },
+
+  getMetric(id) {
+    return this.metricMap.get(id);
+  },
+
+  deleteMetric(id) {
+    return this.metricMap.delete(id);
+  },
+
+  registerMetric(eventName, id, props, type) {
+    if (!this.getMetric(id)) {
+      // hash name and props into table/object
+      const payload = this.createPayload(eventName, props);
+      return this.storeMetric(id, payload, type);
+    }
+    return false;
+  },
+
+  /**
+  * Registers a metric to be used to later update or call finish method
+  * @param {string} eventName name of the metric
+  * @param {string} id unique identifier
+  * @param {Object} props to populate the metric with
+  * @returns {bool} true if metric registered, false otherwise
+  */
+  clientRegisterMetric(eventName, id, props) {
+    return this.registerMetric(eventName, id, props, this.CLIENT_METRIC);
+  },
+
+  /**
+  * Registers a metric to be used to later update or call finish method
+  * @param {string} eventName name of the metric
+  * @param {string} id unique identifier
+  * @param {Object} props to populate the metric with
+  * @returns {bool} true if metric registered, false otherwise
+  */
+  sdkRegisterMetric(eventName, id, props) {
+    return this.registerMetric(eventName, id, props, this.SDK_METRIC);
+  },
+
+  /**
+   * Updates a registered metric, if it exists
+   * @param {string} id lookup key for metric
+   * @param {Object} metricProps to be used for updating metric
+   * @returns {bool} true if the metrics exists and was updated, false otherwise
+   */
+  updateRegisteredMetric(id, metricProps) {
+    const storedMetric = this.getMetric(id);
+    const props = storedMetric ? storedMetric.metric : null;
+    const type = storedMetric ? storedMetric.type : null;
+
+    if (!props || !type || !metricProps) {
+      return false;
+    }
+
+    // update metric
+    _.assign(props.fields, metricProps.fields);
+
+    return this.storeMetric(id, props, type);
+  },
+
+  metricIsType(id, type) {
+    return this.getMetric(id) && this.getMetric(id).type === type;
+  },
+
+  /**
+   * Sends a registered metric if it exists
+   * @param {string} id of the metric to be sent
+   * @returns {Object} HttpResponse object
+   */
+  clientSendRegisteredMetric(id) {
+    if (!this.metricIsType(id, this.CLIENT_METRIC)) {
+      return false;
+    }
+    const httpResp = this.sendBatchMetric(this.getMetric(id).metric);
+    this.deleteMetric(id);
+    return httpResp;
+  },
+
+  /**
+   * Sends a registered metric if it exists
+   * @param {string} id of the metric to be sent
+   * @returns {Object} HttpResponse object
+   */
+  sdkSendRegisteredMetric(id) {
```

Comment:
So the idea here is that if you are the original creator of the metric, be you the client or the sdk, you are the party that has ultimate authority to send the metric. In order to otherwise know the role of the party, we would have to pass a param to the method.

Using a wrapper like this ensures that we are not unintentionally sending a metric that is registered and owned by the client in the SDK, or vice versa. 

I'm not sure I understand the argument for only having a single send metric method. Are you suggesting a paramaterized version where we send the identifier as an argument, (e.g.: "SDK" or "Client")?

So that would look like sendMetric(id, requestingParty){
if(getMetric(id).type != requestingParty) {
// party does not own this metric, cannot send, only update!
return false
}
else {
send metric
}

Reply:
i was still thinking there'd be two send methods: `sendRegisteredMetric()` and `sendRegisteredClientMetric()`. `sendRegisteredMetric()` should be a noop if `registerClientMetric()` is called for that key before `registerMetric()` is called for that key. That way, the sdk can report a given dataset without client input via 

- sdk: `registerMetric()`
- sdk: `updateMetric()` 
- sdk: `sendMetric()`

should a client choose to participate, the flow would be 

- client: `registerClientMetric()`
- sdk: `registerMetric()`
- sdk: `updateMetric()` 
- sdk: `sendMetric()` (noop in this case)
- client: `sendClientMetric()`

Reply:
done